Problem with migrations location when switched to 0.4.3

After switching to 0.4.3 from 0.4.2 starter does not see migration script directory in packaged boot jar. I.e. when one clean up elastic and then start migration via boot jar app startup. Everything worked fine in 0.4.2 version

Well, actually I cannot use evolution version 0.4.3 with reflections 0.9.12:
Caused by: java.lang.NoClassDefFoundError: org/reflections/scanners/Scanners
        at com.senacor.elasticsearch.evolution.core.internal.migration.input.MigrationScriptReaderImpl.readScriptsFromClassPath(MigrationScriptReaderImpl.java:150)

Running just main class is ok, running via gradle bootRun is ok, the problem appears only if I run bootJar via java -jar
So, I think there is a problem scanning bootJar classpath and receiving resources (migrations) from BOOT_INF/classes directory.
I.e. some mechanism changed in reflections library, that leads to error.

--- a/README.md
+++ b/README.md
@@ -291,9 +291,10 @@ ElasticsearchEvolution.configure()
 
 ## 6 changelog
 
-### v0.4.4-SNAPSHOT
+### v0.5.0-SNAPSHOT
 
-- ...
+- added spring boot configuration metadata [#240](https://github.com/senacor/elasticsearch-evolution/pull/240)
+- replaces unmaintained [org.reflections](https://github.com/ronmamo/reflections) library with [classgraph](https://github.com/classgraph/classgraph) to scan the classpath for migration files. Fixes [#239](https://github.com/senacor/elasticsearch-evolution/issues/239) 
 
 ### v0.4.3
 
@@ -304,6 +305,10 @@ ElasticsearchEvolution.configure()
 - added regression tests on JDK 21
 - added regression tests for spring boot 3.1
 - update org.reflections:reflections from 0.9.12 to 0.10.2 [#233](https://github.com/senacor/elasticsearch-evolution/pull/233) thanks @RiVogel
+- **KNOWN ISSUES**:
+  - [#239](https://github.com/senacor/elasticsearch-evolution/issues/239): Migration files not found in Spring Boot jar
+    - Workaround 1: downgrade `org.reflections:reflections` to `0.10.1`
+    - Workaround 2: downgrade `elasticsearch-evolution` to `0.4.2`
 
 ### v0.4.2
 

--- a/elasticsearch-evolution-core/pom.xml
+++ b/elasticsearch-evolution-core/pom.xml
@@ -6,7 +6,7 @@
     <parent>
         <groupId>com.senacor.elasticsearch.evolution</groupId>
         <artifactId>elasticsearch-evolution-parent</artifactId>
-        <version>0.4.4-SNAPSHOT</version>
+        <version>0.5.0-SNAPSHOT</version>
         <relativePath>../</relativePath>
     </parent>
     <artifactId>elasticsearch-evolution-core</artifactId>
@@ -50,8 +50,8 @@
         </dependency>
 
         <dependency>
-            <groupId>org.reflections</groupId>
-            <artifactId>reflections</artifactId>
+            <groupId>io.github.classgraph</groupId>
+            <artifactId>classgraph</artifactId>
         </dependency>
 
         <dependency>

--- a/elasticsearch-evolution-core/src/main/java/com/senacor/elasticsearch/evolution/core/api/config/ElasticsearchEvolutionConfig.java
+++ b/elasticsearch-evolution-core/src/main/java/com/senacor/elasticsearch/evolution/core/api/config/ElasticsearchEvolutionConfig.java
@@ -27,6 +27,9 @@ public class ElasticsearchEvolutionConfig {
 
     /**
      * Locations of migrations scripts. Supported is classpath:some/path and file:/some/path
+     * The location is scanned recursive.
+     * NOTE: all scripts in all locations / subdirectories will be flatted and only the version number will be used to
+     * order them.
      */
     private List<String> locations = new ArrayList<>(
             Collections.singletonList("classpath:es/migration"));
@@ -281,14 +284,6 @@ public class ElasticsearchEvolutionConfig {
     public ElasticsearchEvolutionConfig setHistoryMaxQuerySize(int historyMaxQuerySize) {
         this.historyMaxQuerySize = historyMaxQuerySize;
         return this;
-    }
-
-    /**
-     * @deprecated use {@link #isValidateOnMigrate()} instead
-     */
-    @Deprecated
-    public boolean getValidateOnMigrate() {
-        return isValidateOnMigrate();
     }
 
     public boolean isValidateOnMigrate() {

--- a/elasticsearch-evolution-core/src/main/java/com/senacor/elasticsearch/evolution/core/internal/migration/input/MigrationScriptReaderImpl.java
+++ b/elasticsearch-evolution-core/src/main/java/com/senacor/elasticsearch/evolution/core/internal/migration/input/MigrationScriptReaderImpl.java
@@ -3,33 +3,25 @@ package com.senacor.elasticsearch.evolution.core.internal.migration.input;
 import com.senacor.elasticsearch.evolution.core.api.MigrationException;
 import com.senacor.elasticsearch.evolution.core.api.migration.MigrationScriptReader;
 import com.senacor.elasticsearch.evolution.core.internal.model.migration.RawMigrationScript;
-import java.util.regex.Pattern;
-import org.reflections.Reflections;
-import org.reflections.scanners.Scanners;
-import org.reflections.util.ClasspathHelper;
-import org.reflections.util.ConfigurationBuilder;
-import org.reflections.util.FilterBuilder;
+import io.github.classgraph.ClassGraph;
+import io.github.classgraph.ScanResult;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.BufferedReader;
+import java.io.ByteArrayInputStream;
 import java.io.IOException;
-import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.net.URI;
 import java.net.URISyntaxException;
-import java.net.URL;
 import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
-import java.util.Collection;
+import java.util.ArrayList;
 import java.util.List;
-import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
-
-import static java.util.Collections.emptySet;
 
 /**
  * @author Andreas Keefer
@@ -137,35 +129,26 @@ public class MigrationScriptReaderImpl implements MigrationScriptReader {
             // otherwise e.g. "...location_some_suffix" will also be found when search for "...location".
             location = location + "/";
         }
-        final String locationWithoutPrefixAsPackageNotation = location.substring(CLASSPATH_PREFIX.length())
-                .replace("/", ".");
 
-        final Collection<URL> urls = ClasspathHelper.forPackage(locationWithoutPrefixAsPackageNotation);
-        final Set<String> resources;
-        if (urls.isEmpty()) {
-            // https://github.com/senacor/elasticsearch-evolution/issues/27
-            // when the package is empty or does not exist, Reflections can't find any URLs to scan for
-            resources = emptySet();
-        } else {
-            Reflections reflections = new Reflections(new ConfigurationBuilder()
-                    .setScanners(Scanners.Resources)
-                    .filterInputsBy(new FilterBuilder().includePackage(locationWithoutPrefixAsPackageNotation))
-                    .setUrls(urls));
-            resources = reflections.getResources(Pattern.compile(esMigrationPrefix + ".*"))
-                    .stream()
-                    .filter(path -> isValidFilename(Paths.get(path).getFileName().toString()))
-                    .collect(Collectors.toSet());
+        final String locationWithoutPrefix = location.substring(CLASSPATH_PREFIX.length());
+
+        List<RawMigrationScript> res = new ArrayList<>();
+        try (ScanResult scanResult = new ClassGraph()
+                .acceptPaths(locationWithoutPrefix)
+                .scan()) {
+            scanResult.getAllResources()
+                    .filter(resource -> isValidFilename(Paths.get(resource.getPath()).getFileName().toString()))
+                    .forEach(resource -> {
+                        logger.debug("reading migration script '{}' from classpath...", resource);
+                        try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(new ByteArrayInputStream(resource.load()), encoding))) {
+                            Path p = Paths.get(resource.getPath());
+                            res.addAll(read(bufferedReader, p.getFileName().toString()).collect(Collectors.toList()));
+                        } catch (IOException e) {
+                            throw new MigrationException("can't read script from classpath: " + resource, e);
+                        }
+                    });
         }
-
-        return resources.stream().flatMap(resource -> {
-            logger.debug("reading migration script '{}' from classpath...", resource);
-            try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(getInputStream(resource), encoding))) {
-                Path p = Paths.get(resource);
-                return read(bufferedReader, p.getFileName().toString());
-            } catch (IOException e) {
-                throw new MigrationException("can't read script from classpath: " + resource, e);
-            }
-        });
+        return res.stream();
     }
 
     private Stream<RawMigrationScript> read(BufferedReader reader, String filename) {
@@ -178,44 +161,10 @@ public class MigrationScriptReaderImpl implements MigrationScriptReader {
         return Stream.of(new RawMigrationScript().setFileName(filename).setContent(content));
     }
 
-    public static InputStream getInputStream(String path) {
-        ClassLoader classLoader = getDefaultClassLoader();
-        if (classLoader != null) {
-            return classLoader.getResourceAsStream(path);
-        } else {
-            return ClassLoader.getSystemResourceAsStream(path);
-        }
-    }
-
     private boolean hasValidSuffix(String path) {
         return this.esMigrationSuffixes
                 .stream()
                 .anyMatch(suffix -> path.toLowerCase().endsWith(suffix.toLowerCase()));
-    }
-
-    static ClassLoader getDefaultClassLoader() {
-        ClassLoader cl = null;
-        try {
-            cl = Thread.currentThread().getContextClassLoader();
-            logger.trace("getDefaultClassLoader - Thread.currentThread().getContextClassLoader()='{}'", cl);
-        } catch (Throwable ex) {
-            // Cannot access thread context ClassLoader - falling back...
-        }
-        if (cl == null) {
-            // No thread context class loader -> use class loader of this class.
-            cl = MigrationScriptReaderImpl.class.getClassLoader();
-            logger.trace("getDefaultClassLoader - MigrationScriptReaderImpl.class.getClassLoader()='{}'", cl);
-            if (cl == null) {
-                // getClassLoader() returning null indicates the bootstrap ClassLoader
-                try {
-                    cl = ClassLoader.getSystemClassLoader();
-                    logger.trace("getDefaultClassLoader - ClassLoader.getSystemClassLoader()='{}'", cl);
-                } catch (Throwable ex) {
-                    // Cannot access system ClassLoader - oh well, maybe the caller can live with null...
-                }
-            }
-        }
-        return cl;
     }
 
     private boolean isValidFilename(String fileName) {

--- a/elasticsearch-evolution-core/src/test/java/com/senacor/elasticsearch/evolution/core/api/config/ElasticsearchEvolutionConfigTest.java
+++ b/elasticsearch-evolution-core/src/test/java/com/senacor/elasticsearch/evolution/core/api/config/ElasticsearchEvolutionConfigTest.java
@@ -1,18 +1,11 @@
 package com.senacor.elasticsearch.evolution.core.api.config;
 
-import com.senacor.elasticsearch.evolution.core.internal.migration.input.MigrationScriptReaderImpl;
+import io.github.classgraph.ClassGraph;
+import io.github.classgraph.ScanResult;
 import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
-import org.reflections.Reflections;
-import org.reflections.scanners.Scanners;
-import org.reflections.util.ClasspathHelper;
-import org.reflections.util.ConfigurationBuilder;
-import org.reflections.util.FilterBuilder;
 
-import java.io.BufferedReader;
 import java.io.IOException;
-import java.io.InputStreamReader;
-import java.nio.charset.StandardCharsets;
 import java.util.Collections;
 import java.util.List;
 import java.util.stream.Collectors;
@@ -27,21 +20,19 @@ class ElasticsearchEvolutionConfigTest {
 
     @Test
     void springConfigurationMmetadataJson_should_have_been_generated_by_springBootConfigurationProcessor() {
-        final String metadataPath = "META-INF";
-        Reflections reflections = new Reflections(new ConfigurationBuilder()
-                .setScanners(Scanners.Resources)
-                .filterInputsBy(new FilterBuilder().includePackage(metadataPath))
-                .setUrls(ClasspathHelper.forPackage(metadataPath)));
-        final List<String> springConfigMetadata = reflections.getResources("spring-configuration-metadata\\.json")
-                .stream()
-                .map(resource -> {
-                    try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(MigrationScriptReaderImpl.getInputStream(resource), StandardCharsets.UTF_8))) {
-                        return bufferedReader.lines()
-                                .collect(Collectors.joining("\n"));
-                    } catch (IOException e) {
-                        throw new IllegalStateException("can't read spring-configuration-metadata.json from classpath: " + resource, e);
-                    }
-                }).collect(Collectors.toList());
+        final List<String> springConfigMetadata;
+        try (ScanResult scanResult = new ClassGraph().acceptPathsNonRecursive("META-INF/").scan()) {
+            springConfigMetadata = scanResult.getResourcesWithLeafName("spring-configuration-metadata.json")
+                    .stream()
+                    .map(resource -> {
+                        try {
+                            return resource.getContentAsString();
+                        } catch (IOException e) {
+                            throw new IllegalStateException(e);
+                        }
+                    })
+                    .collect(Collectors.toList());
+        }
 
         assertThat(springConfigMetadata)
                 .anySatisfy(metadataContent -> assertThat(metadataContent).contains(ElasticsearchEvolutionConfig.class.getName()));

--- a/elasticsearch-evolution-core/src/test/java/com/senacor/elasticsearch/evolution/core/internal/migration/input/MigrationScriptReaderImplTest.java
+++ b/elasticsearch-evolution-core/src/test/java/com/senacor/elasticsearch/evolution/core/internal/migration/input/MigrationScriptReaderImplTest.java
@@ -50,8 +50,10 @@ class MigrationScriptReaderImplTest {
 
                 List<RawMigrationScript> res = reader.read();
 
-                assertThat(res).hasSize(1);
-                assertThat(res.get(0).getFileName()).isEqualTo("MANIFEST.MF");
+                assertThat(res).isNotEmpty()
+                        .allSatisfy(rawMigrationScript -> assertThat(rawMigrationScript.getFileName())
+                                .as("fileName")
+                                .isEqualTo("MANIFEST.MF"));
             }
 
             @Test
@@ -248,11 +250,33 @@ class MigrationScriptReaderImplTest {
     }
 
     private URL resolveURL(String path) {
-        ClassLoader classLoader = MigrationScriptReaderImpl.getDefaultClassLoader();
+        ClassLoader classLoader = getDefaultClassLoader();
         if (classLoader != null) {
             return classLoader.getResource(path);
         } else {
             return ClassLoader.getSystemResource(path);
         }
+    }
+
+    static ClassLoader getDefaultClassLoader() {
+        ClassLoader cl = null;
+        try {
+            cl = Thread.currentThread().getContextClassLoader();
+        } catch (Throwable ex) {
+            // Cannot access thread context ClassLoader - falling back...
+        }
+        if (cl == null) {
+            // No thread context class loader -> use class loader of this class.
+            cl = MigrationScriptReaderImpl.class.getClassLoader();
+            if (cl == null) {
+                // getClassLoader() returning null indicates the bootstrap ClassLoader
+                try {
+                    cl = ClassLoader.getSystemClassLoader();
+                } catch (Throwable ex) {
+                    // Cannot access system ClassLoader - oh well, maybe the caller can live with null...
+                }
+            }
+        }
+        return cl;
     }
 }

--- a/pom.xml
+++ b/pom.xml
@@ -5,7 +5,7 @@
 
     <groupId>com.senacor.elasticsearch.evolution</groupId>
     <artifactId>elasticsearch-evolution-parent</artifactId>
-    <version>0.4.4-SNAPSHOT</version>
+    <version>0.5.0-SNAPSHOT</version>
     <packaging>pom</packaging>
     <parent>
         <groupId>org.springframework.boot</groupId>
@@ -102,7 +102,7 @@
         <commons-io.version>2.15.0</commons-io.version>
         <!--when updating elasticsearch versions, also update "ElasticsearchContainer" version in EmbeddedElasticsearchExtension-->
         <elasticsearch.version>7.5.2</elasticsearch.version>
-        <reflections.version>0.10.2</reflections.version>
+        <classgraph.version>4.8.164</classgraph.version>
         <testcontainers.elasticsearch.version>1.19.2</testcontainers.elasticsearch.version>
         <lombok.version>1.18.30</lombok.version>
         <!-- newer byte-buddy version to support JDK 21 -->
@@ -112,9 +112,9 @@
     <dependencyManagement>
         <dependencies>
             <dependency>
-                <groupId>org.reflections</groupId>
-                <artifactId>reflections</artifactId>
-                <version>${reflections.version}</version>
+                <groupId>io.github.classgraph</groupId>
+                <artifactId>classgraph</artifactId>
+                <version>${classgraph.version}</version>
             </dependency>
 
             <!-- Elasticsearch -->

--- a/spring-boot-starter-elasticsearch-evolution/pom.xml
+++ b/spring-boot-starter-elasticsearch-evolution/pom.xml
@@ -6,7 +6,7 @@
     <parent>
         <groupId>com.senacor.elasticsearch.evolution</groupId>
         <artifactId>elasticsearch-evolution-parent</artifactId>
-        <version>0.4.4-SNAPSHOT</version>
+        <version>0.5.0-SNAPSHOT</version>
         <relativePath>../</relativePath>
     </parent>
     <artifactId>spring-boot-starter-elasticsearch-evolution</artifactId>

--- a/tests/migration-scripts/pom.xml
+++ b/tests/migration-scripts/pom.xml
@@ -4,7 +4,7 @@
     <modelVersion>4.0.0</modelVersion>
     <groupId>com.senacor.elasticsearch.evolution</groupId>
     <artifactId>migration-scripts</artifactId>
-    <version>0.4.4-SNAPSHOT</version>
+    <version>0.5.0-SNAPSHOT</version>
     <description>jar containing migration files</description>
 
     <packaging>jar</packaging>

--- a/tests/pom.xml
+++ b/tests/pom.xml
@@ -6,7 +6,7 @@
     <parent>
         <groupId>com.senacor.elasticsearch.evolution</groupId>
         <artifactId>elasticsearch-evolution-parent</artifactId>
-        <version>0.4.4-SNAPSHOT</version>
+        <version>0.5.0-SNAPSHOT</version>
         <relativePath>../</relativePath>
     </parent>
     <artifactId>tests</artifactId>

--- a/tests/test-spring-boot-2.1-scriptsInJarFile/pom.xml
+++ b/tests/test-spring-boot-2.1-scriptsInJarFile/pom.xml
@@ -10,7 +10,7 @@
     </parent>
     <groupId>com.senacor.elasticsearch.evolution</groupId>
     <artifactId>test-spring-boot-2.1-scriptsInJarFile</artifactId>
-    <version>0.4.4-SNAPSHOT</version>
+    <version>0.5.0-SNAPSHOT</version>
     <description>Demo project for Spring Boot</description>
 
     <properties>

--- a/tests/test-spring-boot-2.2/pom.xml
+++ b/tests/test-spring-boot-2.2/pom.xml
@@ -10,7 +10,7 @@
 	</parent>
 	<groupId>com.senacor.elasticsearch.evolution</groupId>
 	<artifactId>test-spring-boot-2.2</artifactId>
-	<version>0.4.4-SNAPSHOT</version>
+	<version>0.5.0-SNAPSHOT</version>
 	<description>Demo project for Spring Boot</description>
 
 	<properties>

--- a/tests/test-spring-boot-2.3/pom.xml
+++ b/tests/test-spring-boot-2.3/pom.xml
@@ -10,7 +10,7 @@
 	</parent>
 	<groupId>com.senacor.elasticsearch.evolution</groupId>
 	<artifactId>test-spring-boot-2.3</artifactId>
-	<version>0.4.4-SNAPSHOT</version>
+	<version>0.5.0-SNAPSHOT</version>
 	<description>Demo project for Spring Boot</description>
 
 	<properties>

--- a/tests/test-spring-boot-2.4/pom.xml
+++ b/tests/test-spring-boot-2.4/pom.xml
@@ -10,7 +10,7 @@
 	</parent>
 	<groupId>com.senacor.elasticsearch.evolution</groupId>
 	<artifactId>test-spring-boot-2.4</artifactId>
-	<version>0.4.4-SNAPSHOT</version>
+	<version>0.5.0-SNAPSHOT</version>
 	<description>Demo project for Spring Boot</description>
 
 	<properties>

--- a/tests/test-spring-boot-2.5-scriptsInJarFile/pom.xml
+++ b/tests/test-spring-boot-2.5-scriptsInJarFile/pom.xml
@@ -10,7 +10,7 @@
     </parent>
     <groupId>com.senacor.elasticsearch.evolution</groupId>
     <artifactId>test-spring-boot-2.5-scriptsInJarFile</artifactId>
-    <version>0.4.4-SNAPSHOT</version>
+    <version>0.5.0-SNAPSHOT</version>
     <description>Demo project for Spring Boot</description>
 
     <properties>

--- a/tests/test-spring-boot-2.6/pom.xml
+++ b/tests/test-spring-boot-2.6/pom.xml
@@ -10,7 +10,7 @@
 	</parent>
 	<groupId>com.senacor.elasticsearch.evolution</groupId>
 	<artifactId>test-spring-boot-2.6</artifactId>
-	<version>0.4.4-SNAPSHOT</version>
+	<version>0.5.0-SNAPSHOT</version>
 	<description>Demo project for Spring Boot</description>
 
 	<properties>

--- a/tests/test-spring-boot-2.7/pom.xml
+++ b/tests/test-spring-boot-2.7/pom.xml
@@ -10,7 +10,7 @@
 	</parent>
 	<groupId>com.senacor.elasticsearch.evolution</groupId>
 	<artifactId>test-spring-boot-2.7</artifactId>
-	<version>0.4.4-SNAPSHOT</version>
+	<version>0.5.0-SNAPSHOT</version>
 	<description>Demo project for Spring Boot</description>
 
 	<properties>

--- a/tests/test-spring-boot-3.0-scriptsInJarFile/pom.xml
+++ b/tests/test-spring-boot-3.0-scriptsInJarFile/pom.xml
@@ -10,7 +10,7 @@
     </parent>
     <groupId>com.senacor.elasticsearch.evolution</groupId>
     <artifactId>test-spring-boot-3.0-scriptsInJarFile</artifactId>
-    <version>0.4.4-SNAPSHOT</version>
+    <version>0.5.0-SNAPSHOT</version>
     <description>Demo project for Spring Boot</description>
 
     <properties>

--- a/tests/test-spring-boot-3.1/pom.xml
+++ b/tests/test-spring-boot-3.1/pom.xml
@@ -10,7 +10,7 @@
 	</parent>
 	<groupId>com.senacor.elasticsearch.evolution</groupId>
 	<artifactId>test-spring-boot-3.1</artifactId>
-	<version>0.4.4-SNAPSHOT</version>
+	<version>0.5.0-SNAPSHOT</version>
 	<description>Demo project for Spring Boot</description>
 
 	<properties>

